Link basics yubo

yubo56 opened the pull request:
Implemented Link and related subclasses/events. Notably, LinkSendEvent and DeviceToLinkEvent should be working.

Records timing of pushed LinkBufferElements for possible future dynamic deque of wait times, but currently not implemented.

yubo56 committed:
First commit of Link, should have most from design doc

Few TODOS remaining:
    run() for events needs global time, then can implement time_entry on
        LinkBufferElement
    Implement average wait time for last N packets

yubo56 committed:
Removed \'new\' keyword in a few places, oops

yubo56 committed:
test

yubo56 committed:
Merge branch 'master' network_simulator into link-basics-yubo

New main_event_loop run()

yubo56 committed:
Global time in run() is implemented

Everything basic about Link should work. TODO are RouterReceivedPacket
signature and deque of wait times

sharonyang commented:
I think you can get rid of all the "old mode" -> "new mode" issues
based on what was discussed here: http://stackoverflow.com/questions/1257592/how-do-i-remove-files-saying-old-mode-100755-new-mode-100644-from-unstaged-cha

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
-        :ivar int buffer_size_bits: maximum buffer size in bits.
+        :ivar int max_buffer_size_bits: maximum buffer size in bits.
         :ivar Queue queue: a FIFO queue of LinkBufferElements
         """
```
:ivar for curr_buffer_size_bits.

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
-        self.buffer_size_bits = buffer_size_bits
+        self.max_buffer_size_bits = max_buffer_size_bits
         self.queue = queue
+        self.curr_buffer_size_bits = 0 # size of current buffer in bits
```
Remove this and use the :ivar above.

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
+        """
+        if isinstance(packet, RouterPacket):
+            # always push RouterPacket
+            self.queue.put(LinkBufferElement(packet, dest_device_addr,\
```
nit: awk spacing. usually I'd do:
self.queue.put(blahblahblah(
    things, things, things, things))

or
self.queue.put(blahblahblah(things,
                                             things,
                                             things,
                                             things))

sharonyang replied:
the bottom three "things" were meant to align with the first "things"
lol

yubo56 replied:
done! for reasons

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
+        """
+        return self.max_buffer_size_bits - self.curr_buffer_size_bits
+
+    def push(self, packet, dest_device_addr, global_clock_sec):
```
I thought we can retrieve dest_device_addr from the packet?

yubo56 replied:
oops I'm dumb. Fixing all

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
+        :param string dest_device_addr: destination device of packet
+
+        Pushes packet onto LinkBuffer going towards dest_device_addr
+        :return: True if pushed, False if not
```
nit: comment style for a function:
"""
Purpose of this function.

:param type param_name: what this param is.
:ivar type ivar_name: what is ivar is.
<:return:>
"""

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
+        Pushes packet onto LinkBuffer going towards dest_device_addr
+        :return: True if pushed, False if not
+        """
+        if isinstance(packet, RouterPacket):
```
No size check here?

yubo56 replied:
I decided that RouterPackets are always pushed and don't contribute to statistics, so no size check since always pushed. Is that correct/reasonable?

sharonyang replied:
@lakshbhasin check on this.

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
+            self.curr_buffer_size_bits += DATA_PACKET_SIZE_BITS
+            self.queue.put(LinkBufferElement(packet, dest_device_addr,\
+                    global_clock_sec))
+            return True;
```
nit: ";" not necessary

yubo56 replied:
never knew, it's like I'm writing in Python or something </s>

I'm so silly. done

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
+            self.curr_buffer_size_bits += ACK_PACKET_SIZE_BITS
+            self.queue.put(LinkBufferElement(packet, dest_device_addr,\
+                    global_clock_sec))
+            return True;
```
nit: ";" not necessary

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
+                    global_clock_sec))
+            return True;
+        else:
+            return False; # failed to push
```
nit: ";" not necessary

sharonyang commented on network_simulator/link.py:
```diff
@@ -4,44 +4,94 @@
     """
-
-    def __init__(self, packet, dest_device_addr):
+    def __init__(self, packet, dest_device_addr, entry_time):
```
above: packet has dest_device_addr already?

sharonyang replied:
default value necessary?

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -50,6 +100,20 @@ def __eq__(self, other):
+        return packet of LinkBufferElement
+        :return Packet packet that is in the buffer element
+        """
+        return packet
```
return self.packet.

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -50,6 +100,20 @@ def __eq__(self, other):
+        return destination of LinkBufferElement
+        :return string destination of the buffer element
+        """
+        return dest_device_addr
```
return self.dest_device_addr.

yubo56 replied:
done >.<

sharonyang commented on network_simulator/link.py:
```diff
@@ -95,3 +159,117 @@ def __eq__(self, other):
+
+    def push(self, packet, dest_device_addr, global_clock_sec):
+        """
+        :param Packet packet: Packet to push
```
Comment style. Description goes first. See above.

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -95,3 +159,117 @@ def __eq__(self, other):
+    def __init__(self, link):
+        """
+        :ivar link: link that must send its next packet
+        """
```
Describe ivar buffer_elem.

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -95,3 +159,117 @@ def __eq__(self, other):
+        main_event_loop.schedule_event_with_delay(RouterReceivedPacketEvent(0),\
+                self.link.static_delay_sec + self.packet.size_bits / \
+                    self.link.capacity_bps)
+            # TODO get RouterReceivedPacketEvent
```
TODO(name): Task description.

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -95,3 +159,117 @@ def __eq__(self, other):
+        """
+        self.link = link
+        self.buffer_elem = None
+    def run(self, main_event_loop, statistics):
```
Space between functions (1). Spaces between classes (3).

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -95,3 +159,117 @@ def __eq__(self, other):
+
+        # sanity check
+        if(self.link.queue.empty()):
+            logger.warning("LinkSendEvent while Link empty!")
```
Do we have a logger class imported? Does this compile?
LOL sorry I don't know if we do.

yubo56 replied:
I tried to copy whatever was in main_event_loop, network_topology. I forgot to getLogger() at the top, but fixed that now. Hopefully it works?

Wait, how do I check if it compiles :o

sharonyang replied:
When you click "Run" on PyCharm, it should exit without error.

lakshbhasin replied:
nit: I would use logger.error here just so we can check the log for errors more easily.

sharonyang commented on network_simulator/link.py:
```diff
@@ -95,3 +159,117 @@ def __eq__(self, other):
+        """
+        :ivar packet: packet arriving at Link
+        :ivar link: link that packet is arriving at
+        :ivar dest_device_addr: destination of packet
```
again. packet with dest device addr?

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -95,3 +159,117 @@ def __eq__(self, other):
+            statistics.link_buffer_occ_change(self, self.packet, \
+                    main_event_loop.global_clock_sec) 
+        else:
+            statistics.link_packet_loss(self, main_event_loop.global_clock_sec)
```
No need to pass in self.

yubo56 replied:
Sorry, maybe I don'tk now python, but link_packet_loss requires (self, link, clock), so I'm trying to pass self as argument to the link parameter?

sharonyang replied:
So self is a thing in Python to make the implementation easier in a class.
http://stackoverflow.com/questions/2709821/what-is-the-purpose-of-self-in-python
Without that said, what is needed is link here, so it should be ...loss(self.link, main_event_loop...)

yubo56 replied:
done

sharonyang commented on network_simulator/link.py:
```diff
@@ -95,3 +159,117 @@ def __eq__(self, other):
+        # if successfully pushed
+        if(self.link.push(self.packet, self.dest_device_addr,\
+                main_event_loop.global_cloc_sec)):
+            statistics.link_buffer_occ_change(self, self.packet, \
```
No need to pass in self.

sharonyang replied:
link_buffer_occ_change(self.link, main_event_loop.global_clock_sec)

yubo56 committed:
Fixed Sharon's code review comments

yubo56 committed:
Fixed Sharon's code review changes

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -4,44 +4,98 @@
-        :ivar int buffer_size_bits: maximum buffer size in bits.
+        :ivar int max_buffer_size_bits: maximum buffer size in bits.
         :ivar Queue queue: a FIFO queue of LinkBufferElements
+        :ivar int curr_buffer_size_bits: current buffer size in bits
```
You need to initialize self.curr_buffer_size_bits to 0 inside this constructor.

yubo56 replied:
done

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -4,44 +4,98 @@
     def __repr__(self):
         return str(self.__dict__)
 
+    def remaining(self):
```
Maybe rename to "remaining_size_bits". This can be done by selecting "remaining" and pressing Shift+F6 in PyCharm.

yubo56 replied:
done

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -4,44 +4,98 @@
+        if isinstance(packet, RouterPacket):
+            # always push RouterPacket
+            self.queue.put(LinkBufferElement(\
+                    packet, packet.dest_id, global_clock_sec))
```
The LinkBufferElement constructor only takes in (packet, entry_time). So all the constructor calls should be fixed.

yubo56 replied:
oops, done

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -4,44 +4,98 @@
+            return True
+
+        elif isinstance(packet, DataPacket) and \
+                self.remaining() > DATA_PACKET_SIZE_BITS:
```
You can combine the DataPacket and AckPacket cases in this conditional by just using packet.size_bits instead of using constants like DATA_PACKET_SIZE_BITS etc. Since both AckPackets and DataPackets are treated the same with respect to pushing.

yubo56 replied:
done

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -4,44 +4,98 @@
+        :return: top LinkBufferElement
+        """
+        return self.queue.get()
+
```
nit: two white lines between classes (and between functions outside of a class).

yubo56 replied:
oops, thought sharon said 2, fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -4,44 +4,98 @@
 
 class LinkBufferElement(object):
     """
     Wrapper class containing a Packet and which Device address it's going to
```
*contains a Packet and its entry time.

Also, you need to know the direction that this packet is going (is it going to the Device on one end or the Device on the other). Otherwise, you can't send it to the right place. See comments in LinkSendEvent below.

yubo56 replied:
done

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -4,44 +4,98 @@
-        :ivar Packet packet
-        :ivar string dest_device_addr
+        :ivar Packet packet packet in buffer element
+        :ivar entry_time current time that packet is being pushed onto queue
```
nit: reword as "time when Packet is added to buffer"

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -50,6 +104,21 @@ def __eq__(self, other):
         return isinstance(other, self.__class__) and \
             self.__dict__ == other.__dict__
 
+    def get_packet(self):
```
Getters really aren't necessary for most Python stuff, since Python doesn't have private variables. This isn't a huge style issue, it'll just make parts of the codebase inconsistent.

Edit: actually, are these being used? I don't think they are.

yubo56 replied:
They aren't being used, just wanted to put them in in case we wanted to them at some point. Took out for now

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+        :return: float of link delay time in seconds
+        """
+
+        # TODO consider average wait time for last N packets
```
(comment doesn't require response) just wanted to note that we will have to add queuing delay at some point (since if we didn't routing table updates wouldn't be that useful). I don't think it's extra credit.

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+
+class LinkSendEvent(Event):
+    """
+    Event representing a packet arriving at the Link
```
This represents a Packet leaving the Link's buffer.

yubo56 replied:
oops fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+    def __init__(self, link):
+        """
+        :ivar link: link that must send its next packet
+        :ivar buff_elem: temporary variable containing LinkBufferElement to send
```
Maybe just initialize the self.buff_elem to None in the constructor so the comments match the constructor code.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+        self.buffer_elem = self.link.pop()
+
+        # update statistics only if not RouterPacket
+        if(not isinstance(self.buffer_elem.packet, RouterPacket)):
```
nit: no parentheses around what's inside the if in Python. Just do "if not isinstance... :"

yubo56 replied:
oops, fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+
+        # update statistics only if not RouterPacket
+        if(not isinstance(self.buffer_elem.packet, RouterPacket)):
+            statistics.link_packet_transmitted(link, buffer_elem.packet, \
```
self.link and self.buffer_elem.packet

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+        be scheduled.
+        """
+        # schedule a RouterReceivedPacket with propagation + transmission delay
+        main_event_loop.schedule_event_with_delay(RouterReceivedPacketEvent(0),\
```
You have to send either a RouterReceivedPacketEvent or a HostReceivedPacketEvent, depending on whether the "other end" of the link (which is opposite to where the Packet currently is) is a Host or a Router. This is why you need to know which Device the Packet is going towards.

This is why LinkBufferElements need to include the Device they're going towards.

yubo56 replied:
Fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+        # schedule a RouterReceivedPacket with propagation + transmission delay
+        main_event_loop.schedule_event_with_delay(RouterReceivedPacketEvent(0),\
+                self.link.static_delay_sec + self.packet.size_bits / \
+                    self.link.capacity_bps)
```
It's always important to make sure you're not doing integer division in Python by accident. I would always do float() on either the numerator or denominator to be sure.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+        else:
+            # queue new event with just transmission delay
+            main_event_loop.schedule_event_with_delay(LinkSendEvent(self.link),\
+                    self.buffer_elem.packet.size_bits / self.link.capacity_bps)
```
Ditto integer division comment.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+            main_event_loop.schedule_event_with_delay(LinkSendEvent(self.link),\
+                    self.buffer_elem.packet.size_bits / self.link.capacity_bps)
+
+
```
nit: ditto two lines of whitespace between classes.

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+
+class DeviceToLinkEvent(Event):
+    """
+    Event representing a packet arriving at the Link
```
This actually represents a Packet going from a Device to one of its connected Links' buffers. Which is more verbose but more accurate.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+    def __init__(self, packet, link):
+        """
+        :ivar packet: packet arriving at Link
+        :ivar link: link that packet is arriving at
```
Again, you need to know the direction this Packet is going in (is it going to end_1_device or end_2_device). This needs to be added to the LinkBufferElement

yubo56 replied:
Added

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+        :param Statistics statistics: the Statistics to update
+        """
+        # if successfully pushed
+        if(self.link.push(self.packet, main_event_loop.global_cloc_sec)):
```
no parentheses after "if". Ditto below.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -4,44 +4,98 @@
+        """
+        if isinstance(packet, RouterPacket):
+            # always push RouterPacket
+            self.queue.put(LinkBufferElement(\
```
LinkBufferElement needs to include destination Device as mentioned below. Also, since the code for the LinkBufferElement you construct doesn't depend on the packet type, just write one line before the if instead of multiple.

yubo56 replied:
Done

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        if(not self.link.busy): # only schedule new event if link not busy;
```
This is a strange commenting format. Just put the comments on the lines before the "if", starting with a "#" on each new line.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedules new Events. This is called immediately after run().
```
Make comments for overriding methods more specifics.

yubo56 replied:
fixed? Not 100% sure what you meant by this but I tried

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+            # if link_buffer.queue is empty, then link is no longer busy
+            self.link.busy = False
+        else:
+            # queue new event with just transmission delay
```
nit: maybe assert "self.link.busy == True" in this else clause.

yubo56 committed:
Fixed Laksh's Code Review

Removed link.end_1_addr, since can just do link.end_1_device.address
Added dest_dev to LinkBufferElement and percolated changes, so now when
queueing LinkSendEvent we know whether to generate Router, Host
ReceivedPacketEvent.

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -4,44 +4,86 @@
     """
-    Wrapper class containing a Packet and which Device address it's going to
-    (i.e. the direction).
+    Wrapper class containing a Packet and time of entry to buffer
```
nit: and the Device the Packet is going to on the Link.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -59,28 +101,24 @@ class Link(object):
+            link_buffer=LinkBuffer(), static_delay_sec=None,
+            capacity_bps=None, busy = False):
         """
-        :ivar string end_1_addr: address of Device on one end (e.g. "H1").
```
Don't get rid of end_1_addr and end_2_addr. I need them for reading in from JSON, since we cannot read in references to Devices from JSON. See the NetworkTopology code for how this setup is done.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +133,133 @@ def __eq__(self, other):
+                # TODO(Laksh): RouterReceivedPacketEvent signature
+                # recommended signature RouterReceivedPacketEvent(router, packet)?
+        else:
+            main_event_loop.schedule_event_with_delay(\
```
maybe "assert isinstance(self.buffer_elem.dest_dev, Host)" inside this else.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +133,133 @@ def __eq__(self, other):
+            self.link.busy = False
+        else:
+            # queue new event with just transmission delay
+            assert(self.link.busy == True)
```
nit: no need for parentheses around the inside of an assert call. "assert" is not a function in Python.

yubo56 replied:
fixed sorry I'm dumb

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +133,133 @@ def __eq__(self, other):
+        """
+        :ivar packet: packet arriving at Link
+        :ivar link: link that packet is arriving at
+        :ivar dest_dev: destination Device of the packet pushed onto link
```
Add that this must be one of the two Devices on the ends of the Link.

lakshbhasin commented on network_simulator/link.py:
```diff
@@ -95,3 +133,133 @@ def __eq__(self, other):
+        """
+        self.packet = packet
+        self.link = link
+        self.dest_dev = dest_dev
```
Add assertion that either link.end_1_device == dest_dev or link.end_2_device == dest_dev, just as a check for now.

yubo56 committed:
Fixed Laksh's Code Review

Restored end_1_addr stuff, added sanity checks
